pyroute2/wiset.py

```diff
@@ -0,0 +1,477 @@
+'''
+WiSet module
+============
+
+High level ipset support.
+
+When IPSet module is providing a direct netlink socket with low level
+functions, a WiSet object is built to map ipset objects from kernel. It
+helps to add/remove entries, list content, etc.
+
+For example, adding an entry with IPSet object implies to set a various
+number of parameters:
+
+>>> ipset = IPSet()
+>>> ipset.add("foo", "1.2.3.4/24", etype="net")
+>>> ipset.close()
+
+When they are discovered by a Wiset:
+
+>>> wiset = list_ipset("foo")
+>>> wiset.add("1.2.3.4/24")
+
+Listing entries is also easier using WiSet, since it parses for you
+netlink messages:
+
+>>> wiset.content
+{'1.2.3.0/24': IPStats(packets=None, bytes=None, comment=None, timeout=None)}
+'''
+
+import errno
+import uuid
+from collections import namedtuple
+from inspect import getcallargs
+from socket import AF_INET
+
+from pyroute2 import IPSet
+from pyroute2.netlink.exceptions import IPSetError
+from pyroute2.netlink.nfnetlink.ipset import IPSET_FLAG_WITH_COUNTERS
+from pyroute2.netlink.nfnetlink.ipset import IPSET_FLAG_WITH_COMMENT
+
+
+# Debug variable to detect netlink socket leaks
+COUNT = {"count": 0}
+
+
+def need_ipset_socket(fun):
+    """ Decorator to create netlink socket if needed.
```

Comment:
In the future it may be reasonable to move it to the common module and use not only for ipset.

Reply:
Yes, I will move it when merging the routing/ip part.

Otherwise I updated the pull request with other comment and I added documentation support (it adds a small conflict on index.rst, easy to solve)